gsl/gsl_byte

```diff
@@ -36,8 +36,167 @@
 
 #endif // _MSC_VER
 
+#define GSL_BYTE_TYPE_ENUM 0
+#define GSL_BYTE_TYPE_UCHAR 1
+#define GSL_BYTE_TYPE_STRUCT 2
+
+#ifndef GSL_USE_BYTE_TYPE
+// currently clang++ as well as g++ don't allow a pointer to
+//
+// < enum class byte : unsigned char >
+//
+// to alias other types.
+// so we have to use a typedef to unsigned char instead.
+#ifdef _MSC_VER
+    #define GSL_USE_BYTE_TYPE GSL_BYTE_TYPE_ENUM
+#else
+    #define GSL_USE_BYTE_TYPE GSL_BYTE_TYPE_ENUM
+#endif
+
+#endif
+
 namespace gsl
 {
+//################################### implemet byte as unsigned char ####################################
+#if GSL_USE_BYTE_TYPE == GSL_BYTE_TYPE_UCHAR
+
+using byte = unsigned char;
+
+template <class IntegerType, class = std::enable_if_t<std::is_integral<IntegerType>::value>>
+inline constexpr IntegerType to_integer(byte b) noexcept
+{
+    return{ b };
+}
+
+template<bool E, typename T>
+inline constexpr byte to_byte_impl(T t) noexcept
+{
+    static_assert(
+        E,
+        "gsl::to_byte(t) must be provided an unsigned char, otherwise data loss may occur. "
+        "If you are calling to_byte with an integer contant use: gsl::to_byte<t>() version."
+        );
+    return static_cast<byte>(t);
+}
+
+template<>
+inline constexpr byte to_byte_impl<true, unsigned char>(unsigned char t) noexcept
+{
+    return byte(t);
+}
+
+template<typename T>
+inline constexpr byte to_byte(T t) noexcept
+{
+    return to_byte_impl<std::is_same<T, unsigned char>::value, T>(t);
+}
+
+template <int I>
+inline constexpr byte to_byte() noexcept
+{
+    static_assert(I >= 0 && I <= 255, "gsl::byte only has 8 bits of storage, values must be in range 0-255");
+    return static_cast<byte>(I);
+}
+
+//################################### implemet byte as struct ####################################
+
+#elif GSL_USE_BYTE_TYPE == GSL_BYTE_TYPE_STRUCT
+
+// This is a simple definition for now that allows
+// use of byte within span<> to be standards-compliant
+struct byte {
+    unsigned char v;
+    explicit operator unsigned char&() { return v; }
+    explicit operator const unsigned char&() const { return v; }
+    friend bool operator==(byte l, byte r) { return l.v == r.v; }
+    friend bool operator!=(byte l, byte r) { return l.v != r.v; }
+};
+
+template <class IntegerType, class = std::enable_if_t<std::is_integral<IntegerType>::value>>
+inline constexpr byte& operator<<=(byte& b, IntegerType shift) noexcept
+{
+    return b.v <<= shift, b;
```

Comment:
no, I just called `make test`  and assumed that would call all tests. But yes, even after explicitly selecting the test as shown by you, it passes.